.github/workflows/bench.yml

```diff
@@ -0,0 +1,54 @@
+name: Benchmarks
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+  workflow_dispatch:
+
+jobs:
+  bench:
+    runs-on: [self-hosted, "${{ matrix.os }}", "${{ matrix.device }}"]
+
+    permissions:
+      contents: write
+      deployments: write
+
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [Linux, macOS]
+        feature: [default]
+        device: [cpu]
+        include:
+          - os: Linux
+            feature: cuda
+            device: nvidia_rtx_a5000
+          - os: macOS
+            feature: metal
+            device: intel_uhd_630
+
+    env:
+      FEATURE: ${{ matrix.feature }}
+
+    steps:
+      - uses: actions/checkout@v3
+      - uses: risc0/actions-rs-toolchain@v1
+        with:
+          toolchain: stable
+      - run: cargo bench -F $FEATURE --bench fib -- --output-format=bencher | tee output.txt
+      - name: Store benchmark result
+        uses: risc0/github-action-benchmark@7eeba8924d4ed651010d72d7319b064fc0c10355
```

Comment:
Let's pin to a specific version unless you expect certain behavior on a non-released commit?

Reply:
Github recommends commit hash over version for actions generally to prevent non-determinism but since we own the forks I am happy to swap to a version here.